### PR TITLE
QUIC: CID conformance

### PR DESCRIPTION
**Pull request by paulidale**
- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Review by t8m — CHANGES_REQUESTED** [review]

File: ssl/quic/quic_channel.c [hunk 1]
```diff
@@ -2164,22 +2164,61 @@ void ossl_quic_channel_on_new_conn_id(QUIC_CHANNEL *ch,
...
+     * RFC 9000-5.1.1: An endpoint MUST NOT provide more connection IDs
+     * than the peer's limit.
+     *
+     * After processing a NEW_CONNECTION_ID frame and adding and retiring
+     * active connection IDs, if the number of active connection IDs exceeds
+     * the value advertised in its active_connection_id_limit transport
+     * parameter, an endpoint MUST close the connection with an error of
+     * type CONNECTION_ID_LIMIT_ERROR.
+     */
+    if (new_remote_seq_num - new_retire_prior_to > 1) {
```
[the remark this answers is not in the document]
Comment (paulidale) [reply ssl/quic/quic_channel.c:2194]: It's going to need reworking once we properly support connections.  I'd prefer to leave until then.

**Review by t8m — APPROVED** [review]

File: ssl/quic/quic_channel.c [hunk 2]
```diff
@@ -2164,22 +2164,61 @@ void ossl_quic_channel_on_new_conn_id(QUIC_CHANNEL *ch,
...
+        if (f->seq_num != new_remote_seq_num + 1) {
+            /*
+             * RFC 9000-5.1.1: The sequence number on each newly issued
+             * connection ID MUST increase by 1
+             */
+            ossl_quic_channel_raise_protocol_error(ch,
+                                                   QUIC_ERR_PROTOCOL_VIOLATION,
+                                                   OSSL_QUIC_FRAME_TYPE_NEW_CONN_ID,
+                                                   "sequence number increased by > 1");
+            return;
```
[the remark this answers is not in the document]
Comment (paulidale) [reply ssl/quic/quic_channel.c]: I just removed the check -- we can't validate out of order packet seq_no increase by one IMO.
We also don't send new CID packets, so there is nothing needed for this.

**Comment by paulidale** [comment]
Ping @openssl/committers for second review please.

File: ssl/quic/quic_channel.c [hunk 3]
```diff
@@ -2169,17 +2169,44 @@ void ossl_quic_channel_on_new_conn_id(QUIC_CHANNEL *ch,
...
+     * RFC 9000-5.1.2: An endpoint SHOULD allow for sending and tracking
+     * a number of RETIRE_CONNECTION_ID frames of at least twice the value
+     * of the active_connection_id_limit transport parameter.  An endpoint
+     * MUST NOT forget a connection ID without retiring it, though it MAY
+     * choose to treat having connection IDs in need of retirement that
+     * exceed this limit as a connection error of type CONNECTION_ID_LIMIT_ERROR.
+     *
+     * We are a little bit more liberal than the minimum mandated.
+     */
+    if (new_retire_prior_to - ch->cur_retire_prior_to > 10) {
```
[the remark this answers is not in the document]
Comment (t8m) [reply ssl/quic/quic_channel.c:2205]: Yes, it is, see the code above.

**Comment by paulidale** [comment]
Merged, thanks for the comments

**Pull request closed without merging**